File: machineLearning/classifier/python/train.py
import os
import numpy as np
import cv2
import time

labels = []
trainingFilenames = []
imgs = []
root = "training_files/"

def gets_training_files():
    """
    Pega todas os nomes das aquivos em training_files
    """
    for classe in os.listdir(root):    
        for img in os.listdir(root+str(classe)):
            trainingFilenames.append(root+str(classe)+"/"+img)
            labels.append(classe)

def SVMParams(svm):
    print("Parametros da SVM:\n\
           Kernel type     :{}\n\
           Type            :{}\n\
           C               :{}\n\
           Degree          :{}\n\
           Nu              :{}\n\
           Gamma           :{}\n\
    ".format(svm.getKernelType(), svm.getType(), svm.getC(), svm.getDegree(), svm.getNu(), svm.getGamma()))


gets_training_files()

i = time.clock()
print("\nAnalizando rotulos....")
for index, file in enumerate(trainingFilenames):    
    #print("Analizando rotulo (Classe|Arquivo)-> {}|{}".format(labels[index], trainingFilenames[index]))
    img = cv2.imread(trainingFilenames[index], 0)
    imgs.append(img.flatten())
imgs = np.array(imgs, dtype=np.float32)
labels = np.array(labels, dtype=np.int32)
f = time.clock()
print("A analise dos rotulos levou: {}s".format(f-i))


#Configura a SVM
#Seta os melhores parametros 
svm = cv2.ml.SVM_create()
svm.setType(cv2.ml.SVM_C_SVC)
svm.setKernel(cv2.ml.SVM_POLY)
svm.setC(1)
svm.setGamma(0.50625)
svm.setDegree(3)

i = time.clock()
print("Treinando o classificador.......")
svm.train(imgs, cv2.ml.ROW_SAMPLE, labels)
f = time.clock()
print("O treinamento levou: {}s".format(f-i))

#Salva a SVM
print("Salvando a SVM.......")
svm.save("rats_and_pitons.yml")
SVMParams(svm)

File: machineLearning/classifier/cpp/train.cpp
// Bibliotecas Padrao
#include <iostream>
#include <vector>

// Testes do tempo de execucao
#include <ctime>

// OpenCV
#include <opencv2/core.hpp>
#include <opencv2/ml.hpp>

using namespace std;
using namespace cv;

//Mostra os parametros da SVM
void SVMParams(ml::SVM *svm){
    cout << "Kernel type     : " << svm->getKernelType() << endl;
    cout << "Type            : " << svm->getType() << endl;
    cout << "C               : " << svm->getC() << endl;
    cout << "Degree          : " << svm->getDegree() << endl;
    cout << "Nu              : " << svm->getNu() << endl;
    cout << "Gamma           : " << svm->getGamma() << endl;
}

//Todas as ações da SVM
void SVMtrain(Mat &trainMat, Mat &labelsMat, string &metodo){    
    // Configura a SVM
    //Kernel do tipo polinomial
    
    Ptr<ml::SVM> svm = cv::ml::SVM::create();
    svm->setKernel(ml::SVM::POLY);
	svm->setType(ml::SVM::C_SVC);
    svm->setC(1);
    svm->setDegree(0.6);
    svm->setGamma(0.50625);
    
    /*
    //Kernel do tipo RBF 
    Ptr<ml::SVM> svm = ml::SVM::create();
    svm->setType(ml::SVM::C_SVC);
    svm->setKernel(ml::SVM::RBF);
    svm->setGamma(0.50625);
    svm->setC(1);
    */
    /*
    //kernel do tipo CHI2
    Ptr<ml::SVM> svm = ml::SVM::create();
    svm->setType(ml::SVM::C_SVC);
    svm->setKernel(ml::SVM::CHI2);
    //parametros do auto train
    svm->setC(1);
    svm->setGamma(1e-5);
    */
    /*
    //kernel do tipo SIGMOID
    Ptr<ml::SVM> svm = ml::SVM::create();
    svm->setType(ml::SVM::C_SVC);
    svm->setKernel(ml::SVM::SIGMOID);
    */
    //Treina o classificador 
    int i = clock();
    cout << "Treinando o classificador ...\n";
    svm->train(trainMat, ml::ROW_SAMPLE, labelsMat);
    //svm->trainAuto(traiMat, ml::ROW_SAMPLE, labelsMat);
    int f = clock();
    cout << "O treinamento levou: " << (f-i)/(float)CLOCKS_PER_SEC << "s\n";

    // Salva a SVM
    cout << "Salvando a SVM ...\n";
    svm->save("ymls/"+ metodo +"/RATS_POLY.yml");
    cout <<  endl;
    SVMParams(svm);
}


int main(int argc, char **argv){
    cout << "******Classificacao de Imagens******\n";

    string metodo = argv[1];
    
    cout << "Lendo os dados ...\n";
    FileStorage fs("ymls/"+ metodo +"/trainMat.yml", FileStorage::READ);
    Mat trainMat, labelsMat;
    fs["labelsMat"] >> labelsMat;
    fs["trainMat"] >> trainMat;
    fs.release(); 

    SVMtrain(trainMat, labelsMat, metodo);

    return 0;    
}